### PR TITLE
Add event filter for instrumentation-device CG reconciler

### PR DESCRIPTION
**Pull request by RonFed**
During the un-instrumentation of a large number of sources`reconcileSingleWorkload` may be called concurrently by the collectors group reconciler and the instrumented-application reconciler.

This race condition can cause the instrumentation device to be added right after its removal as demonstrated in the following log:

```
2024-11-08T12:18:23Z	INFO	removed instrumentation device from workload	{"controller": "instrumentationdevice-instrumentedapplication", "controllerGroup": "odigos.io", "controllerKind": "InstrumentedApplication", "InstrumentedApplication": {"name":"deployment-coupon","namespace":"simple-demo18"}, "namespace": "simple-demo18", "name": "deployment-coupon", "reconcileID": "a50d0c85-fd07-4896-84a3-817472fc3953", "namespace": "simple-demo18", "kind": "&TypeMeta{Kind:Deployment,APIVersion:apps/v1,}", "name": "coupon", "reason": "NoRuntimeDetails"}
2024-11-08T12:18:23Z	DEBUG	instrumented application deleted	{"controller": "deleteinstrumentedapplication-deployment", "controllerGroup": "apps", "controllerKind": "Deployment", "Deployment": {"name":"coupon","namespace":"simple-demo18"}, "namespace": "simple-demo18", "name": "coupon", "reconcileID": "08560386-4960-4ace-aaae-343fd6fb4cc2", "namespace": "simple-demo18", "name": "coupon", "kind": "Deployment"}
2024-11-08T12:18:23Z	INFO	added instrumentation device to workload	{"controller": "instrumentationdevice-collectorsgroup", "controllerGroup": "odigos.io", "controllerKind": "CollectorsGroup", "CollectorsGroup": {"name":"odigos-data-collection","namespace":"odigos-system"}, "namespace": "odigos-system", "name": "odigos-data-collection", "reconcileID": "52989ae7-2686-4e1c-99cb-23b3252535ad", "name": "coupon", "namespace": "simple-demo18"}
```

To address this, an event filter is added to the CG reconciler, which only forward either
1. Create events which have CG ready.
2. Update events where the CG became ready.

In addition, the reconciler is updated to get a fresh copy from the cache of each instrumented app it handles before it reconciles it. This should reduce the chance of handling an old instrumented application and reconciling workloads unnecessarily.

### CODE REVIEW

**Review by edeNFed — APPROVED** [review]

File: instrumentor/controllers/instrumentationdevice/collectorsgroup_controller.go [hunk 1]
```diff
@@ -18,34 +18,116 @@ package instrumentationdevice
...
+
+		if err != nil {
+			reconcileErr = errors.Join(reconcileErr, err)
+			continue
+		}
+
+		err = reconcileSingleWorkload(ctx, r.Client, &currentInstApp, isDataCollectionReady)
 		if err != nil {
-			return ctrl.Result{}, err
+			reconcileErr = errors.Join(reconcileErr, err)
```
Comment (blumamir) [inline instrumentor/controllers/instrumentationdevice/collectorsgroup_controller.go:128]: I wonder - is joining them one by one creates some hierarchy in the result, vs collecting them into an array and Joining the array at one go?

**Review by blumamir — APPROVED** [review]

File: instrumentor/controllers/instrumentationdevice/collectorsgroup_controller.go [hunk 2]
```diff
@@ -18,34 +18,116 @@ package instrumentationdevice
...
+// once the collectors group becomes ready - by adding the instrumentation device to the workloads.
+// This is necessary to ensure that we won't instrument any workload before the
+// node collectors are ready to receive the data.
 type CollectorsGroupReconciler struct {
 	client.Client
 	Scheme *runtime.Scheme
 }
 
-func (r *CollectorsGroupReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+type collectorsGroupReadyPredicate struct {
```
Comment (blumamir) [inline instrumentor/controllers/instrumentationdevice/collectorsgroup_controller.go:45]: notice that in https://github.com/odigos-io/odigos/pull/1714 I extracted this predicate into `k8sutils` so you can refactor this part if we merge it first

File: instrumentor/controllers/instrumentationdevice/collectorsgroup_controller.go (hunk 1 above)
Comment (RonFed) [reply instrumentor/controllers/instrumentationdevice/collectorsgroup_controller.go:128]: https://go.dev/play/p/dTpcbqrWT7h

Looks to be identical

File: instrumentor/controllers/instrumentationdevice/collectorsgroup_controller.go (hunk 2 above)
Comment (RonFed) [reply instrumentor/controllers/instrumentationdevice/collectorsgroup_controller.go]: Done

**Pull request merged**